sound/soc/sof/intel/telemetry.c

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+//
+
+#include <sound/sof/ipc4/header.h>
+#include <sound/sof/xtensa.h>
+#include "../ipc4-priv.h"
+#include "../sof-priv.h"
+#include "hda.h"
+#include "telemetry.h"
+
+#define CHECK_SEPARATOR(S, V) \
+	if (S != V) { \
+		dev_err(sdev->dev, "error: %s : %#x is not matched with %#x", #S, S, V); \
+		return false; \
+	} \
+
+static u32 find_telemetry_slots(struct snd_sof_dev *sdev)
+{
+	u32 slot_desc_type_offset;
+	u32 type;
+	int i;
+
+	for (i = 0; i < SOF_IPC4_MAX_DEBUG_SLOTS; i++) {
+		/* The type is the second u32 in the slot descriptor */
+		slot_desc_type_offset = sdev->debug_box.offset;
+		slot_desc_type_offset += SOF_IPC4_DEBUG_DESCRIPTOR_SIZE * i + sizeof(u32);
```

Comment:
how about `slot_desc_type_offset = sdev->debug_box.offset + sizeof(u32);` before the loop and then `slot_desc_type_offset += SOF_IPC4_DEBUG_DESCRIPTOR_SIZE;` at the end of each loop iteration

Reply:
sure. manual optimization